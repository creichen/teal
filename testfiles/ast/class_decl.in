class MyClass = {}

class MyClass [MyType1, MyType2] = {}

class MyClass (arg1, arg2) = {}

class MyClass <: Super1, Super2 = {}

class MyClass [MyType1, MyType2] (arg1, args2) = {}

class MyClass [MyType1, MyType2] <: Super1, Super2 [MyType3, MyType4] = {}

class MyClass (arg1, arg2) <: Super1, Super2 [MyType3, MyType4] = {}



class MyClass = {
      // variables
      v : MyType;
      w : T := v;

      // methods
      fun f(a) = return a;

      // do we want statemens inside a class? they don't work for now
      // x := 5;
}