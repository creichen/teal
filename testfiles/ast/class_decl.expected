Program
  List
  List
    ClassDecl
      IdDecl ID="MyClass"
      List
      List
      List
      List
    ClassDecl
      IdDecl ID="MyClass"
      List
        TypeVarDecl
          IdDecl ID="MyType1"
        TypeVarDecl
          IdDecl ID="MyType2"
      List
      List
      List
    ClassDecl
      IdDecl ID="MyClass"
      List
      List
        VarDecl
          IdDecl ID="arg1"
          Opt
          Opt
        VarDecl
          IdDecl ID="arg2"
          Opt
          Opt
      List
      List
    ClassDecl
      IdDecl ID="MyClass"
      List
      List
      List
        SimpleType
          IdUse ID="Super1"
          List
        SimpleType
          IdUse ID="Super2"
          List
      List
    ClassDecl
      IdDecl ID="MyClass"
      List
        TypeVarDecl
          IdDecl ID="MyType1"
        TypeVarDecl
          IdDecl ID="MyType2"
      List
        VarDecl
          IdDecl ID="arg1"
          Opt
          Opt
        VarDecl
          IdDecl ID="args2"
          Opt
          Opt
      List
      List
    ClassDecl
      IdDecl ID="MyClass"
      List
        TypeVarDecl
          IdDecl ID="MyType1"
        TypeVarDecl
          IdDecl ID="MyType2"
      List
      List
        SimpleType
          IdUse ID="Super1"
          List
        SimpleType
          IdUse ID="Super2"
          List
            SimpleType
              IdUse ID="MyType3"
              List
            SimpleType
              IdUse ID="MyType4"
              List
      List
    ClassDecl
      IdDecl ID="MyClass"
      List
      List
        VarDecl
          IdDecl ID="arg1"
          Opt
          Opt
        VarDecl
          IdDecl ID="arg2"
          Opt
          Opt
      List
        SimpleType
          IdUse ID="Super1"
          List
        SimpleType
          IdUse ID="Super2"
          List
            SimpleType
              IdUse ID="MyType3"
              List
            SimpleType
              IdUse ID="MyType4"
              List
      List
    ClassDecl
      IdDecl ID="MyClass"
      List
      List
      List
      List
        VarDeclElt
          VarDecl
            IdDecl ID="v"
            Opt
              SimpleType
                IdUse ID="MyType"
                List
            Opt
        VarDeclElt
          VarDecl
            IdDecl ID="w"
            Opt
              SimpleType
                IdUse ID="T"
                List
            Opt
              Access
                IdUse ID="v"
        FunDeclElt
          FunDecl
            IdDecl ID="f"
            Opt
            List
              VarDecl
                IdDecl ID="a"
                Opt
                Opt
            Opt
              ReturnStmt
                Access
                  IdUse ID="a"
